spccore/multipart_upload.py

```diff
@@ -24,7 +44,78 @@ def multipart_upload_file(client,
 
     file_name = os.path.basename(file_path)
     file_size = os.path.getsize(file_path)
-    md5 = get_md5_hex_digest(file_path)
+    md5 = get_md5_hex_digest_for_file(file_path)
+
+    # Step 1: start upload
+    status = _multipart_upload_status(client,
+                                      file_name,
+                                      content_type,
+                                      file_size,
+                                      md5,
+                                      storage_location_id=storage_location_id,
+                                      generate_preview=generate_preview,
+                                      force_restart=force_restart)
+    if 'resultFileHandleId' in status:
+        return int(status['resultFileHandleId'])
+    upload_id = int(status['uploadId'])
+    parts_to_upload = _parts_to_upload(status['partsState'])
+
+    if use_multiple_threads:
+        thread_pool = get_pool()
+    else:
+        thread_pool = get_pool(1)
+
+    def _one_chunk_upload(part: dict):
+        """
+        The function to evoke in each thread
+
+        :param part: an item in BatchPresignedUploadUrlResponse's partPresignedUrls
+        """
+        return _upload_and_add_part(client,
+                                    upload_id,
+                                    part['partNumber'],
+                                    part['uploadPresignedUrl'],
+                                    get_part_data(file_path, int(part['partNumber']), SYNAPSE_DEFAULT_UPLOAD_PART_SIZE))
+
+    # Step 2: upload parts
+    while parts_to_upload and len(parts_to_upload) > 0:
+        pre_signed_urls_generator = _get_batch_pre_signed_url(client, upload_id, content_type, parts_to_upload)
+
+        thread_pool.map(_one_chunk_upload, pre_signed_urls_generator)
```

Comment:
_one_chunk_upload should silently fail on retry-able errors , allowing other chunks to have a chance to be uploaded.